### PR TITLE
HashJoin respects max_joined_block_size_rows

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* HashJoin respects setting `max_joined_block_size_rows` and do not produce large blocks for `ALL JOIN`

Ref https://github.com/ClickHouse/ClickHouse/pull/54662

### CODE REVIEW

**Comment by liuneng1994** [comment]
The code only controls the size of the columns in the left table. In the use case I designed, there will be more columns in the right table. These columns will be generated at once during the join, using a lot of memory. When the right table has more columns (>40), the performance here also has a big problem. This may be the reason why it is slower than my version. The expansion of the right table columns can also cause OOM.  

What is more difficult to handle is that a row of data in the left table may be associated with tens of thousands of records, because the data may be unbalanced. It is necessary to record the processing status when matching the hash map, or do lazy materialization

**Comment by vdimir** [comment]
@liuneng1994 

With current approach we will insert number of rows equals to the number of matches. But it should not be a problem unless it is smaller than block size.

Imagine we have block size `65409` and each row in left table have 500 matches, like in perf test. So we will generate a block `65500` and it should not be an issue. The problem can be when number of matches is more than half of max block size. Do you mean that problem or there's something different?

**Comment by liuneng1994** [comment]
> With current approach we will insert number of rows equals to the number of matches. But it should not be a problem unless it is smaller than block size.

Sorry, I misunderstood, and the memory problem I mentioned has indeed been solved.

File: src/Interpreters/HashJoin.cpp [hunk 1]
```diff
@@ -1592,8 +1621,12 @@ void HashJoin::joinBlockImpl(
 
     bool has_required_right_keys = (required_right_keys.columns() != 0);
     added_columns.need_filter = join_features.need_filter || has_required_right_keys;
+    added_columns.max_joined_block_rows = table_join->maxJoinedBlockRows();
```
Comment (liuneng1994) [inline src/Interpreters/HashJoin.cpp:1624]: Not sure if AddedColumns::columns has pre-allocated memory

**Comment by baibaichen** [comment]
> With current approach we will insert number of rows equals to the number of matches. But it should not be a problem unless it is smaller than block size.
> 
> Imagine we have block size `65409` and each row in left table have 500 matches, like in perf test. So we will generate a block `65500` and it should not be an issue. The problem can be when number of matches is more than half of max block size. Do you mean that problem or there's something different?

@vdimir neng's case doesn‘t include such situation, but his implmentation already solved this issue, i.e., `max_joined_block_size_rows`  will be respected in any case.

As for performance, we guess it would be related with whether memory is pre-allocated or not. see neng's comment on  

```c++
added_columns.max_joined_block_rows = table_join->maxJoinedBlockRows();
```

But, it is just a guess,  probably perf can find something diffierent.

**Comment by zhanglistar** [comment]
@vdimir Any update on this PR? Thanks.

**Comment by vdimir** [comment]
Requested review from random people from the team, maybe someone will take a look.

**Comment by baibaichen** [comment]
> Requested review from random people from the team, maybe someone will take a look.

how's the performance?

**Comment by vdimir** [comment]
> How's the performance compared to https://github.com/ClickHouse/ClickHouse/pull/54662?

Unfortunately a bit worse than in that PR, but it has improvements comparing to `master`. I still cannot wrap my head around what makes crucial difference, because the idea essentially the same. However, I assume it still worth merging this PR first, since it's quite small change and then try to realize if we can make it faster.

**Comment by liuneng1994** [comment]
@vdimir I found the reason for the slowness. After this PR is merged, I will submit a new PR to optimize performance.

**Comment by vdimir** [comment]
> I found the reason for the slowness. After this PR is merged, I will submit a new PR to optimize performance.

Sounds good! (@devcrafter)

**Comment by devcrafter** [comment]
> > I found the reason for the slowness. After this PR is merged, I will submit a new PR to optimize performance.
> 
> Sounds good! (@devcrafter)

@vdimir Looking into it, but please fix the build

**Comment by vdimir** [comment]
CI failures hardly related, but still pretty interesting:


> Stress test (debug) — Check timeout expired [Details](https://s3.amazonaws.com/clickhouse-test-reports/56996/51eec7a185e0ebd15363b4ceb3a981903815ebf1/stress_test__debug_.html)

```
2023.12.19 15:57:34.457840 [ 547 ] {} <Fatal> Application: Child process was terminated by signal 9 (KILL). If it is not done by 'forcestop' command or manually, the possible cause is OOM Killer (see 'dmesg' and look at the '/var/log/kern.log' for the details).
```

> Stress test (tsan) — Hung check failed, possible deadlock found (see hung_check.log) [Details](https://s3.amazonaws.com/clickhouse-test-reports/56996/51eec7a185e0ebd15363b4ceb3a981903815ebf1/stress_test__tsan_.html)

```
2023-12-19 16:48:05,220 Checking if some queries hung
Using queries from '/usr/share/clickhouse-test/queries' directory
Connecting to ClickHouse server... OK
Received exception from server (version 23.12.1):
Code: 219. DB::Exception: Received from localhost:9000. DB::Exception: New table appeared in database being dropped or detached. Try again.. (DATABASE_NOT_EMPTY)
(query: DETACH DATABASE db01802)
Traceback (most recent call last):
  File "/usr/bin/stress", line 374, in <module>
    main()
  File "/usr/bin/stress", line 359, in main
    res = call(cmd, shell=True, stdout=tee.stdin, stderr=STDOUT, timeout=600)
  File "/usr/lib/python3.10/subprocess.py", line 347, in call
    return p.wait(timeout=timeout)
  File "/usr/lib/python3.10/subprocess.py", line 1209, in wait
    return self._wait(timeout=timeout)
  File "/usr/lib/python3.10/subprocess.py", line 1951, in _wait
    raise TimeoutExpired(self.args, timeout)
subprocess.TimeoutExpired: Command '/usr/bin/clickhouse-test --client-option max_untracked_memory=1Gi max_memory_usage_for_user=0 memory_profiler_step=1Gi --database=system --hung-check --report-logs-stats 00001_select_1' timed out after 600 seconds

```

> Upgrade check (msan) — check_status.tsv doesn't exists [Details](https://s3.amazonaws.com/clickhouse-test-reports/56996/51eec7a185e0ebd15363b4ceb3a981903815ebf1/upgrade_check__msan_.html)

<details>
<summary>MemorySanitizer: use-of-uninitialized-value</summary>


```
==231025==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x5646cfdc5fdd in DB::ReplicatedMergeTreeAttachThread::run() build_docker/./src/Storages/MergeTree/ReplicatedMergeTreeAttachThread.cpp:83:9
    #1 0x5646cfdcfc7e in DB::ReplicatedMergeTreeAttachThread::ReplicatedMergeTreeAttachThread(DB::StorageReplicatedMergeTree&)::$_0::operator()() const build_docker/./src/Storages/MergeTree/ReplicatedMergeTreeAttachThread.cpp:24:82
    #2 0x5646cfdcfc7e in decltype(std::declval<DB::ReplicatedMergeTreeAttachThread::ReplicatedMergeTreeAttachThread(DB::StorageReplicatedMergeTree&)::$_0&>()()) std::__1::__invoke[abi:v15000]<DB::ReplicatedMergeTreeAttachThread::ReplicatedMergeTreeAttachThread(DB::StorageReplicatedMergeTree&)::$_0&>(DB::ReplicatedMergeTreeAttachThread::ReplicatedMergeTreeAttachThread(DB::StorageReplicatedMergeTree&)::$_0&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23
    #3 0x5646cfdcfc7e in void std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::ReplicatedMergeTreeAttachThread::ReplicatedMergeTreeAttachThread(DB::StorageReplicatedMergeTree&)::$_0&>(DB::ReplicatedMergeTreeAttachThread::ReplicatedMergeTreeAttachThread(DB::StorageReplicatedMergeTree&)::$_0&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9
...

Member fields were destroyed
    #0 0x56469d600d1d in __sanitizer_dtor_callback_fields (/usr/bin/clickhouse+0x7c86d1d) (BuildId: b02ef26d094f99621fd60a2d973ca865dffde192)
    #1 0x5646cfdc4177 in DB::ReplicatedMergeTreeAttachThread::~ReplicatedMergeTreeAttachThread() build_docker/./src/Storages/MergeTree/ReplicatedMergeTreeAttachThread.h:37:20
    #2 0x5646cfdc4177 in DB::ReplicatedMergeTreeAttachThread::~ReplicatedMergeTreeAttachThread() build_docker/./src/Storages/MergeTree/ReplicatedMergeTreeAttachThread.cpp:32:1
    #3 0x5646ce6f1e5d in std::__1::__optional_destruct_base<DB::ReplicatedMergeTreeAttachThread, false>::~__optional_destruct_base[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/optional:261:21
    #4 0x5646ce6f1e5d in DB::StorageReplicatedMergeTree::~StorageReplicatedMergeTree() build_docker/./src/Storages/StorageReplicatedMergeTree.cpp:5267:1
...
```

https://s3.amazonaws.com/clickhouse-test-reports/56996/51eec7a185e0ebd15363b4ceb3a981903815ebf1/upgrade_check__msan_/stderr.log



</details>

> Upgrade check (tsan) — check_status.tsv doesn't exists [Details](https://s3.amazonaws.com/clickhouse-test-reports/56996/51eec7a185e0ebd15363b4ceb3a981903815ebf1/upgrade_check__tsan_.html)


<details>
<summary>ThreadSanitizer: data race</summary>


```
WARNING: ThreadSanitizer: data race (pid=1018668)
  Read of size 8 at 0x7b7800c6ea28 by main thread:
    #0 std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo>::get[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:801:16 (clickhouse+0x194f74c8) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #1 std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo>::operator bool[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:833:16 (clickhouse+0x194f74c8)
    #2 bool std::__1::operator!=[abi:v15000]<DB::BackgroundSchedulePoolTaskInfo>(std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo> const&, std::nullptr_t) build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:1254:30 (clickhouse+0x194f74c8)
    #3 DB::BackgroundSchedulePoolTaskHolder::operator bool() const build_docker/./src/Core/BackgroundSchedulePool.h:173:55 (clickhouse+0x194f74c8)
    #4 DB::BackgroundJobsAssignee::finish() build_docker/./src/Storages/MergeTree/BackgroundJobsAssignee.cpp:111:9 (clickhouse+0x194f74c8)
    #5 DB::StorageMergeTree::shutdown(bool) build_docker/./src/Storages/StorageMergeTree.cpp:190:36 (clickhouse+0x19a6e56b) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #6 DB::IStorage::flushAndShutdown(bool) build_docker/./src/Storages/IStorage.h:573:9 (clickhouse+0x1726ed4d) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #7 DB::DatabaseWithOwnTablesBase::shutdown() build_docker/./src/Databases/DatabasesCommon.cpp:311:20 (clickhouse+0x1726ed4d)
    #8 DB::DatabaseOnDisk::shutdown() build_docker/./src/Databases/DatabaseOnDisk.cpp:169:32 (clickhouse+0x171f83cf) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #9 DB::DatabaseCatalog::shutdownImpl() build_docker/./src/Interpreters/DatabaseCatalog.cpp:265:26 (clickhouse+0x175a0e20) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #10 DB::DatabaseCatalog::shutdown() build_docker/./src/Interpreters/DatabaseCatalog.cpp:863:27 (clickhouse+0x175a9dab) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #11 DB::ContextSharedPart::shutdown() build_docker/./src/Interpreters/Context.cpp:567:9 (clickhouse+0x174f7495) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #12 DB::Context::shutdown() build_docker/./src/Interpreters/Context.cpp:4172:13 (clickhouse+0x174e6721) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
...

Previous write of size 8 at 0x7b7800c6ea28 by thread T795 (mutexes: write M0):
    #0 std::__1::enable_if<is_move_constructible<DB::BackgroundSchedulePoolTaskInfo*>::value && is_move_assignable<DB::BackgroundSchedulePoolTaskInfo*>::value, void>::type std::__1::swap[abi:v15000]<DB::BackgroundSchedulePoolTaskInfo*>(DB::BackgroundSchedulePoolTaskInfo*&, DB::BackgroundSchedulePoolTaskInfo*&) build_docker/./contrib/llvm-project/libcxx/include/__utility/swap.h:37:7 (clickhouse+0x194f71a8) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #1 std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo>::swap[abi:v15000](std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo>&) build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:761:9 (clickhouse+0x194f71a8)
    #2 std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo>::operator=[abi:v15000](std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo>&&) build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:723:38 (clickhouse+0x194f71a8)
    #3 DB::BackgroundSchedulePoolTaskHolder::operator=(DB::BackgroundSchedulePoolTaskHolder&&) build_docker/./src/Core/BackgroundSchedulePool.h:165:110 (clickhouse+0x194f71a8)
    #4 DB::BackgroundJobsAssignee::start() build_docker/./src/Storages/MergeTree/BackgroundJobsAssignee.cpp:103:16 (clickhouse+0x194f71a8)
    #5 DB::StorageMergeTree::startup() build_docker/./src/Storages/StorageMergeTree.cpp:151:40 (clickhouse+0x19a6e293) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #6 DB::DatabaseOrdinary::startupTableAsync(DB::AsyncLoader&, std::__1::unordered_set<std::__1::shared_ptr<DB::LoadJob>, std::__1::hash<std::__1::shared_ptr<DB::LoadJob>>, std::__1::equal_to<std::__1::shared_ptr<DB::LoadJob>>, std::__1::allocator<std::__1::shared_ptr<DB::LoadJob>>>, DB::QualifiedTableName const&, DB::LoadingStrictnessLevel)::$_0::operator()(DB::AsyncLoader&, std::__1::shared_ptr<DB::LoadJob> const&) const build_docker/./src/Databases/DatabaseOrdinary.cpp:212:24 (clickhouse+0x17222fd2) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #7 decltype(std::declval<DB::DatabaseOrdinary::startupTableAsync
...

```

https://s3.amazonaws.com/clickhouse-test-reports/56996/51eec7a185e0ebd15363b4ceb3a981903815ebf1/upgrade_check__tsan_/stderr.log

</details>

**Comment by vdimir** [comment]
Rebased one more time
@devcrafter

**Review by alexey-milovidov — APPROVED** [review]

File: src/Interpreters/HashJoin.cpp [hunk 2]
```diff
@@ -1365,19 +1385,29 @@ NO_INLINE IColumn::Filter joinRightColumns(
...
         added_columns.offsets_to_replicate = std::make_unique<IColumn::Offsets>(rows);
 
     IColumn::Offset current_offset = 0;
-
-    for (size_t i = 0; i < rows; ++i)
+    size_t max_joined_block_rows = added_columns.max_joined_block_rows;
+    size_t i = 0;
+    for (; i < rows; ++i)
     {
+        if constexpr (join_features.need_replication)
```
Comment (devcrafter) [inline src/Interpreters/HashJoin.cpp:1401]: @vdimir you've explained what `need_replicaion` means at the time, but I've already forgotten and there is no comment. Let's add it next to the `need_replication` definition.

File: src/Interpreters/HashJoin.cpp [hunk 3]
```diff
@@ -1365,19 +1385,29 @@ NO_INLINE IColumn::Filter joinRightColumns(
...
+        {
+            if (unlikely(current_offset > max_joined_block_rows))
+            {
+                added_columns.offsets_to_replicate->resize_assume_reserved(i);
```
Comment (devcrafter) [inline src/Interpreters/HashJoin.cpp:1405]: How do we ensure that `offsets_to_replicate` is not `nullptr` and the size is reserved?

File: src/Interpreters/HashJoin.cpp [hunk 4]
```diff
@@ -1349,19 +1369,29 @@ NO_INLINE IColumn::Filter joinRightColumns(
...
+                added_columns.filter.resize_assume_reserved(i);
```
[the remark this answers is not in the document]
Comment (devcrafter) [reply src/Interpreters/HashJoin.cpp:1406]: We need to ensure that it's reserved

**Review by devcrafter — COMMENTED** [review]
Sorry, I was late

File: src/Interpreters/HashJoin.cpp (hunk 3 above)
Comment (vdimir) [reply src/Interpreters/HashJoin.cpp:1405]: Yes, we initialize it if `join_features.need_replication = true` https://github.com/ClickHouse/ClickHouse/blob/d764d8d72b477470b635106980c0f030caaabba7/src/Interpreters/HashJoin.cpp#L1394-L1395

And use it only when flag is enabled

File: src/Interpreters/HashJoin.cpp [hunk 5]
```diff
@@ -1891,10 +1952,10 @@ HashJoin::~HashJoin()
 {
     if (!data)
     {
-        LOG_TRACE(log, "({}) Join data has been already released", fmt::ptr(this));
+        LOG_TRACE(log, "{}Join data has been already released", instance_log_id);
```
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Interpreters/HashJoin.cpp:1963]: Logger name normally corresponds to component name, but not to particular component instance, imo. So, not sure if we should make logger name dynamic

File: src/Interpreters/HashJoin.cpp (hunk 2 above)
Comment (vdimir) [reply src/Interpreters/HashJoin.cpp:1401]: Added comments and changed `resize_assume_reserved -> resize` https://github.com/ClickHouse/ClickHouse/pull/58289

**Pull request merged**